lorawan: Add application-defined battery level callback API

KubaFYI opened the pull request:
Provides the LoRaWAN stack with an optional callback to be
called whenever the battery level needs to be read.

In the case of callback not being provided, the LoRaWAN
stack will report 255 (battery level unavailable) as per
the LoRaWAN spec.

Signed-off-by: Kuba Sanak <contact@kuba.fyi>

KubaFYI commented:
@Mani-Sadhasivam please let me know if you'd like anything to change with this one.

Mani-Sadhasivam commented on include/lorawan/lorawan.h:
```diff
@@ -99,6 +99,20 @@ struct lorawan_join_config {
+ * Provide the LoRaWAN stack with a function to be called whenever a battery
+ * level needs to be read. The function needs to return a value between 0-254
+ * for battery levels and/or 255 when unable to read the battery or no battery
+ * present.
```
As per the API requirement of loramac-node `0: node is connected to an external power source, 1..254: battery level, where 1 is the minimum and 254 is the maximum value, 255: the node was not able to measure the battery level`

So why do we want to change the behaviour here?

KubaFYI replied:
Fair. Missed the 0 part. Fixed now.

Mani-Sadhasivam commented on include/lorawan/lorawan.h:
```diff
@@ -99,6 +99,24 @@ struct lorawan_join_config {
+ *
+ * @return 0 if successful, negative errno code if failure
+ */
+int lorawan_set_battery_level_callback(uint8_t (*battery_lvl_clbk)(void));
```
Nit: Please stick to `_cb` instead of `_clbk` for callback as it aligns with rest of the Zephyr code.

Mani-Sadhasivam reviewed (changes requested):
Just a minor nit. Otherwise it LGTM!

KubaFYI committed:
lorawan: Add userspace-defined battery level callback API

Provide the LoRaWAN stack with an optional callback to be
called whenever the battery level needs to be read.

In the case of callback not being provided, the LoRaWAN
stack will report 255 (battery level unavailable) as per
the LoRaWAN spec.

Signed-off-by: Kuba Sanak <contact@kuba.fyi>

andrewboie commented:
s/userspace/application/ in your commit title please, this isn't related to user mode threads/CONFIG_USERSPACE

Mani-Sadhasivam commented:
@carlescufi Please merged this one!

carlescufi commented:
> @carlescufi Please merge this one!

Sorry @Mani-Sadhasivam, this one flew under my radar. the change looks good to me and I approved it, but we are now in the stabilization phase for the 2.5 release. This will have to wait until that is complete.
